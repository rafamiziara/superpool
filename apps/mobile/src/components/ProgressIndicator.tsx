import React from 'react';
import { View } from 'react-native';

interface ProgressIndicatorProps {
  totalSteps: number;
  currentStep: number;
  className?: string;
}

export function ProgressIndicator({ totalSteps, currentStep, className = '' }: ProgressIndicatorProps) {
  return (
    <View className={`flex-row justify-center items-center gap-2 ${className}`}>
      {Array.from({ length: totalSteps }, (_, index) => (
        <View
          key={index}
          className={`w-2 h-2 rounded-full ${
            index === currentStep
              ? 'bg-primary'
              : 'bg-muted-foreground/30'
          }`}
        />
      ))}
    </View>
  );
}