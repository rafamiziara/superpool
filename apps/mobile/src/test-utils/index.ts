/**
 * Centralized export of all test utilities
 * Import from this file to get all testing helpers
 */

export * from './mockStores'
export * from './testProviders'
export * from './renderWithStore'

/**
 * Common test utilities and helpers
 */

/**
 * Wait for MobX reactions to settle
 * Useful when testing computed values or reactions
 */
export const waitForMobX = async (timeout = 100): Promise<void> => {
  return new Promise((resolve) => {
    setTimeout(resolve, timeout)
  })
}

/**
 * Mock Firebase Auth user object for testing
 */
export const createMockFirebaseUser = (overrides: Record<string, unknown> = {}) => ({
  uid: 'test-user-id',
  email: 'test@example.com',
  displayName: 'Test User',
  emailVerified: true,
  photoURL: null,
  phoneNumber: null,
  ...overrides,
})

/**
 * Mock Ethereum address for testing
 */
export const MOCK_ETH_ADDRESS = '0x1234567890123456789012345678901234567890'
export const MOCK_ETH_ADDRESS_2 = '0x0987654321098765432109876543210987654321'

/**
 * Common test chain IDs
 */
export const TEST_CHAIN_IDS = {
  ETHEREUM_MAINNET: 1,
  POLYGON_MAINNET: 137,
  POLYGON_AMOY: 80002,
  LOCALHOST: 31337,
} as const
