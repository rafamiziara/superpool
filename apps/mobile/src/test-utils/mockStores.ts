import { AuthenticationStore } from '../stores/AuthenticationStore'
import { WalletStore } from '../stores/WalletStore'
import { PoolManagementStore } from '../stores/PoolManagementStore'
import { UIStore } from '../stores/UIStore'
import { RootStore } from '../stores/RootStore'
import { ErrorType } from '../utils/errorHandling'

/**
 * Factory functions for creating mock MobX stores with preset states
 */

export const createMockAuthenticationStore = (overrides: Partial<AuthenticationStore> = {}): AuthenticationStore => {
  const store = new AuthenticationStore()
  
  // Reset the store to clean state first
  store.completedSteps.clear()
  store.completedSteps.add('connect-wallet') // resetProgress adds this
  
  // Create a mocked version by wrapping the original methods
  const originalStartStep = store.startStep.bind(store)
  const originalCompleteStep = store.completeStep.bind(store) 
  const originalFailStep = store.failStep.bind(store)
  const originalResetProgress = store.resetProgress.bind(store)
  const originalGetStepStatus = store.getStepStatus.bind(store)
  const originalGetStepInfo = store.getStepInfo.bind(store)
  const originalGetAllSteps = store.getAllSteps.bind(store)

  // Create spies that wrap the original functionality
  const mockStartStep = jest.fn().mockImplementation(originalStartStep)
  const mockCompleteStep = jest.fn().mockImplementation(originalCompleteStep)
  const mockFailStep = jest.fn().mockImplementation(originalFailStep)
  const mockResetProgress = jest.fn().mockImplementation(originalResetProgress)
  const mockGetStepStatus = jest.fn().mockImplementation(originalGetStepStatus)
  const mockGetStepInfo = jest.fn().mockImplementation(originalGetStepInfo)
  const mockGetAllSteps = jest.fn().mockImplementation(originalGetAllSteps)

  // Replace methods on the store with mocks
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).startStep = mockStartStep
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).completeStep = mockCompleteStep
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).failStep = mockFailStep
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).resetProgress = mockResetProgress
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).getStepStatus = mockGetStepStatus
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).getStepInfo = mockGetStepInfo
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  ;(store as any).getAllSteps = mockGetAllSteps

  // Apply overrides
  Object.assign(store, overrides)

  return store
}

export const createMockWalletStore = (overrides: Partial<WalletStore> = {}): WalletStore => {
  const store = new WalletStore()

  // Apply overrides
  Object.assign(store, overrides)

  return store
}

export const createMockPoolManagementStore = (overrides: Partial<PoolManagementStore> = {}): PoolManagementStore => {
  const store = new PoolManagementStore()

  // Apply overrides
  Object.assign(store, overrides)

  return store
}

export const createMockUIStore = (overrides: Partial<UIStore> = {}): UIStore => {
  const store = new UIStore()

  // Apply overrides
  Object.assign(store, overrides)

  return store
}

export const createMockRootStore = (
  storeOverrides: {
    authenticationStore?: Partial<AuthenticationStore>
    walletStore?: Partial<WalletStore>
    poolManagementStore?: Partial<PoolManagementStore>
    uiStore?: Partial<UIStore>
  } = {}
): RootStore => {
  const rootStore = new RootStore()

  // Always replace authenticationStore with mocked version for tests
  rootStore.authenticationStore = createMockAuthenticationStore(storeOverrides.authenticationStore || {})

  if (storeOverrides.walletStore) {
    rootStore.walletStore = createMockWalletStore(storeOverrides.walletStore)
  }

  if (storeOverrides.poolManagementStore) {
    rootStore.poolManagementStore = createMockPoolManagementStore(storeOverrides.poolManagementStore)
  }

  if (storeOverrides.uiStore) {
    rootStore.uiStore = createMockUIStore(storeOverrides.uiStore)
  }

  return rootStore
}

/**
 * Preset mock store configurations for common test scenarios
 */

export const mockStorePresets = {
  // Authenticated user with connected wallet
  authenticatedWithWallet: () =>
    createMockRootStore({
      walletStore: {
        isConnected: true,
        address: '0x1234567890123456789012345678901234567890',
        chainId: 137,
        connectionError: null,
        isConnecting: false,
      },
    }),

  // Unauthenticated user
  unauthenticated: () =>
    createMockRootStore({
      walletStore: {
        isConnected: false,
        address: undefined,
        chainId: undefined,
        connectionError: null,
        isConnecting: false,
      },
    }),

  // Authentication in progress
  authenticating: () =>
    createMockRootStore({
      authenticationStore: {
        authLock: {
          isLocked: true,
          startTime: Date.now(),
          walletAddress: '0x1234567890123456789012345678901234567890',
          abortController: new AbortController(),
          requestId: 'test-request-id',
        },
      },
    }),

  // Wallet connection in progress
  connectingWallet: () =>
    createMockRootStore({
      walletStore: {
        isConnected: false,
        address: undefined,
        chainId: undefined,
        connectionError: null,
        isConnecting: true,
      },
    }),

  // Error states
  authenticationError: () =>
    createMockRootStore({
      authenticationStore: {
        authError: {
          name: 'AppError',
          message: 'Authentication failed',
          type: ErrorType.AUTHENTICATION_FAILED,
          userFriendlyMessage: 'Authentication failed',
        },
      },
    }),

  walletConnectionError: () =>
    createMockRootStore({
      walletStore: {
        isConnected: false,
        address: undefined,
        chainId: undefined,
        connectionError: 'Failed to connect wallet',
        isConnecting: false,
      },
    }),
}
