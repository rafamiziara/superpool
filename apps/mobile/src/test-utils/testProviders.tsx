import React, { ReactNode } from 'react'
import { StoreContext } from '../stores/StoreContext'
import { RootStore } from '../stores/RootStore'
import { createMockRootStore } from './mockStores'

/**
 * Test provider that wraps components with a mock MobX store context
 */
interface TestStoreProviderProps {
  children: ReactNode
  store?: RootStore
}

export const TestStoreProvider: React.FC<TestStoreProviderProps> = ({ children, store = createMockRootStore() }) => {
  // Create StoreContext.Provider manually since we need to pass our mock store
  const StoreContextProvider = StoreContext.Provider as React.ComponentType<{
    value: RootStore
    children: ReactNode
  }>

  return <StoreContextProvider value={store}>{children}</StoreContextProvider>
}

/**
 * Higher-order component that wraps a component with TestStoreProvider
 */
export const withMockStore = <P extends object>(Component: React.ComponentType<P>, store?: RootStore) => {
  const WrappedComponent: React.FC<P> = (props) => (
    <TestStoreProvider store={store}>
      <Component {...props} />
    </TestStoreProvider>
  )

  WrappedComponent.displayName = `withMockStore(${Component.displayName || Component.name})`

  return WrappedComponent
}
