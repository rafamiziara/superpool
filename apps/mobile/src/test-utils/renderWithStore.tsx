import { render, renderHook, RenderOptions, RenderResult } from '@testing-library/react-native'
import React, { ReactElement } from 'react'
import { RootStore } from '../stores/RootStore'
import { createMockRootStore } from './mockStores'
import { TestStoreProvider } from './testProviders'

/**
 * Custom render options that includes store configuration
 */
interface CustomRenderOptions extends Omit<RenderOptions, 'wrapper'> {
  store?: RootStore
  wrapper?: React.ComponentType<{ children: React.ReactNode }>
}

/**
 * Custom render function that wraps components with TestStoreProvider
 * This is the recommended way to render components that use MobX stores in tests
 */
export const renderWithStore = (ui: ReactElement, options: CustomRenderOptions = {}): RenderResult => {
  const { store = createMockRootStore(), wrapper: CustomWrapper, ...renderOptions } = options

  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    if (CustomWrapper) {
      return (
        <TestStoreProvider store={store}>
          <CustomWrapper>{children}</CustomWrapper>
        </TestStoreProvider>
      )
    }

    return <TestStoreProvider store={store}>{children}</TestStoreProvider>
  }

  return render(ui, { wrapper: Wrapper, ...renderOptions })
}

/**
 * Re-export everything from React Native Testing Library for convenience
 */
export * from '@testing-library/react-native'

/**
 * Override the default render method to always use renderWithStore
 * This ensures all components are rendered with a mock store context
 */
export { renderWithStore as render }

/**
 * Custom renderHook function that wraps hooks with TestStoreProvider
 * This is the recommended way to test hooks that use MobX stores
 */
export const renderHookWithStore = <TResult,>(hook: () => TResult, options: { store?: RootStore } = {}) => {
  const { store = createMockRootStore() } = options

  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => {
    return <TestStoreProvider store={store}>{children}</TestStoreProvider>
  }

  return renderHook(() => hook(), { wrapper: Wrapper })
}

/**
 * Utility function to create a minimal render for non-store components
 * Use this only when you need to test components that don't use MobX stores
 */
export const renderWithoutStore = render as typeof import('@testing-library/react-native').render
