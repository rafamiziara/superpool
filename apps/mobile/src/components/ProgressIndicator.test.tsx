import React from 'react'
import { render } from '../test-utils'
import { ProgressIndicator } from './ProgressIndicator'

describe('ProgressIndicator', () => {
  describe('Basic Rendering', () => {
    it('should render with required props', () => {
      const { getByTestId } = render(<ProgressIndicator totalSteps={3} currentStep={1} />)

      const container = getByTestId('progress-indicator-container')
      expect(container).toBeTruthy()
    })

    it('should render correct number of steps', () => {
      const totalSteps = 5
      const { getAllByTestId } = render(<ProgressIndicator totalSteps={totalSteps} currentStep={2} />)

      const steps = getAllByTestId(/progress-step-\d+/)
      expect(steps).toHaveLength(totalSteps)
    })

    it('should render all step elements with correct testIDs', () => {
      const totalSteps = 4
      const { getByTestId } = render(<ProgressIndicator totalSteps={totalSteps} currentStep={1} />)

      for (let i = 0; i < totalSteps; i++) {
        const step = getByTestId(`progress-step-${i}`)
        expect(step).toBeTruthy()
      }
    })
  })

  describe('Step State Visualization', () => {
    it('should highlight current step correctly', () => {
      const currentStep = 2
      const { getByTestId } = render(<ProgressIndicator totalSteps={5} currentStep={currentStep} />)

      const currentStepElement = getByTestId(`progress-step-${currentStep}`)
      expect(currentStepElement.props.className).toContain('bg-primary')
    })

    it('should style non-current steps correctly', () => {
      const currentStep = 2
      const totalSteps = 5
      const { getByTestId } = render(<ProgressIndicator totalSteps={totalSteps} currentStep={currentStep} />)

      for (let i = 0; i < totalSteps; i++) {
        const stepElement = getByTestId(`progress-step-${i}`)
        if (i !== currentStep) {
          expect(stepElement.props.className).toContain('bg-muted-foreground/30')
          expect(stepElement.props.className).not.toContain('bg-primary')
        }
      }
    })

    it('should handle first step as current', () => {
      const { getByTestId } = render(<ProgressIndicator totalSteps={3} currentStep={0} />)

      const firstStep = getByTestId('progress-step-0')
      expect(firstStep.props.className).toContain('bg-primary')

      const secondStep = getByTestId('progress-step-1')
      expect(secondStep.props.className).toContain('bg-muted-foreground/30')
    })

    it('should handle last step as current', () => {
      const totalSteps = 4
      const lastStep = totalSteps - 1
      const { getByTestId } = render(<ProgressIndicator totalSteps={totalSteps} currentStep={lastStep} />)

      const lastStepElement = getByTestId(`progress-step-${lastStep}`)
      expect(lastStepElement.props.className).toContain('bg-primary')

      const previousStep = getByTestId(`progress-step-${lastStep - 1}`)
      expect(previousStep.props.className).toContain('bg-muted-foreground/30')
    })
  })

  describe('Edge Cases', () => {
    it('should handle single step', () => {
      const { getByTestId, getAllByTestId } = render(<ProgressIndicator totalSteps={1} currentStep={0} />)

      const steps = getAllByTestId(/progress-step-\d+/)
      expect(steps).toHaveLength(1)

      const singleStep = getByTestId('progress-step-0')
      expect(singleStep.props.className).toContain('bg-primary')
    })

    it('should handle zero steps gracefully', () => {
      const { queryAllByTestId } = render(<ProgressIndicator totalSteps={0} currentStep={0} />)

      const steps = queryAllByTestId(/progress-step-\d+/)
      expect(steps).toHaveLength(0)
    })

    it('should handle currentStep out of range', () => {
      const { getByTestId } = render(<ProgressIndicator totalSteps={3} currentStep={5} />)

      // All steps should be inactive since currentStep is out of range
      for (let i = 0; i < 3; i++) {
        const step = getByTestId(`progress-step-${i}`)
        expect(step.props.className).toContain('bg-muted-foreground/30')
      }
    })

    it('should handle negative currentStep', () => {
      const { getByTestId } = render(<ProgressIndicator totalSteps={3} currentStep={-1} />)

      // All steps should be inactive since currentStep is negative
      for (let i = 0; i < 3; i++) {
        const step = getByTestId(`progress-step-${i}`)
        expect(step.props.className).toContain('bg-muted-foreground/30')
      }
    })
  })

  describe('Styling and Layout', () => {
    it('should apply default className', () => {
      const { getByTestId } = render(<ProgressIndicator totalSteps={3} currentStep={1} />)

      const container = getByTestId('progress-indicator-container')
      expect(container.props.className).toContain('flex-row justify-center items-center gap-2')
    })

    it('should apply custom className', () => {
      const customClassName = 'my-custom-class p-4'
      const { getByTestId } = render(<ProgressIndicator totalSteps={3} currentStep={1} className={customClassName} />)

      const container = getByTestId('progress-indicator-container')
      expect(container.props.className).toContain('flex-row justify-center items-center gap-2')
      expect(container.props.className).toContain(customClassName)
    })

    it('should apply correct step styles', () => {
      const { getByTestId } = render(<ProgressIndicator totalSteps={3} currentStep={1} />)

      for (let i = 0; i < 3; i++) {
        const step = getByTestId(`progress-step-${i}`)
        expect(step.props.className).toContain('w-2 h-2 rounded-full')
      }
    })
  })

  describe('Accessibility', () => {
    it('should have proper accessibility labels for current step', () => {
      const { getByTestId } = render(<ProgressIndicator totalSteps={4} currentStep={2} />)

      const currentStep = getByTestId('progress-step-2')
      expect(currentStep.props.accessibilityLabel).toBe('Step 3 of 4 (current)')
    })

    it('should have proper accessibility labels for completed steps', () => {
      const { getByTestId } = render(<ProgressIndicator totalSteps={5} currentStep={3} />)

      const completedStep = getByTestId('progress-step-1')
      expect(completedStep.props.accessibilityLabel).toBe('Step 2 of 5 (completed)')
    })

    it('should have proper accessibility labels for pending steps', () => {
      const { getByTestId } = render(<ProgressIndicator totalSteps={4} currentStep={1} />)

      const pendingStep = getByTestId('progress-step-3')
      expect(pendingStep.props.accessibilityLabel).toBe('Step 4 of 4 (pending)')
    })

    it('should provide accessibility labels for all steps', () => {
      const totalSteps = 3
      const currentStep = 1
      const { getByTestId } = render(<ProgressIndicator totalSteps={totalSteps} currentStep={currentStep} />)

      for (let i = 0; i < totalSteps; i++) {
        const step = getByTestId(`progress-step-${i}`)
        expect(step.props.accessibilityLabel).toBeTruthy()
        expect(step.props.accessibilityLabel).toContain(`Step ${i + 1} of ${totalSteps}`)
      }
    })
  })

  describe('Dynamic Behavior', () => {
    it('should update when currentStep changes', () => {
      const { getByTestId, rerender } = render(<ProgressIndicator totalSteps={3} currentStep={0} />)

      // Initially first step is active
      expect(getByTestId('progress-step-0').props.className).toContain('bg-primary')
      expect(getByTestId('progress-step-1').props.className).toContain('bg-muted-foreground/30')

      // Rerender with different currentStep
      rerender(<ProgressIndicator totalSteps={3} currentStep={1} />)

      // Now second step should be active
      expect(getByTestId('progress-step-0').props.className).toContain('bg-muted-foreground/30')
      expect(getByTestId('progress-step-1').props.className).toContain('bg-primary')
    })

    it('should update when totalSteps changes', () => {
      const { getAllByTestId, rerender } = render(<ProgressIndicator totalSteps={3} currentStep={1} />)

      expect(getAllByTestId(/progress-step-\d+/)).toHaveLength(3)

      // Rerender with more steps
      rerender(<ProgressIndicator totalSteps={5} currentStep={1} />)

      expect(getAllByTestId(/progress-step-\d+/)).toHaveLength(5)
    })
  })

  describe('Snapshot Testing', () => {
    it('should match snapshot with basic props', () => {
      const component = render(<ProgressIndicator totalSteps={3} currentStep={1} />)
      expect(component.toJSON()).toMatchSnapshot()
    })

    it('should match snapshot with custom className', () => {
      const component = render(<ProgressIndicator totalSteps={5} currentStep={2} className="p-4 bg-gray-100" />)
      expect(component.toJSON()).toMatchSnapshot()
    })

    it('should match snapshot with edge cases', () => {
      const component = render(<ProgressIndicator totalSteps={1} currentStep={0} />)
      expect(component.toJSON()).toMatchSnapshot()
    })
  })
})
