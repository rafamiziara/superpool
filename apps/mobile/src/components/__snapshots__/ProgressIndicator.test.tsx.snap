// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ProgressIndicator Snapshot Testing should match snapshot with basic props 1`] = `
<View
  className="flex-row justify-center items-center gap-2 "
  testID="progress-indicator-container"
>
  <View
    accessibilityLabel="Step 1 of 3 (completed)"
    className="w-2 h-2 rounded-full bg-muted-foreground/30"
    testID="progress-step-0"
  />
  <View
    accessibilityLabel="Step 2 of 3 (current)"
    className="w-2 h-2 rounded-full bg-primary"
    testID="progress-step-1"
  />
  <View
    accessibilityLabel="Step 3 of 3 (pending)"
    className="w-2 h-2 rounded-full bg-muted-foreground/30"
    testID="progress-step-2"
  />
</View>
`;

exports[`ProgressIndicator Snapshot Testing should match snapshot with custom className 1`] = `
<View
  className="flex-row justify-center items-center gap-2 p-4 bg-gray-100"
  testID="progress-indicator-container"
>
  <View
    accessibilityLabel="Step 1 of 5 (completed)"
    className="w-2 h-2 rounded-full bg-muted-foreground/30"
    testID="progress-step-0"
  />
  <View
    accessibilityLabel="Step 2 of 5 (completed)"
    className="w-2 h-2 rounded-full bg-muted-foreground/30"
    testID="progress-step-1"
  />
  <View
    accessibilityLabel="Step 3 of 5 (current)"
    className="w-2 h-2 rounded-full bg-primary"
    testID="progress-step-2"
  />
  <View
    accessibilityLabel="Step 4 of 5 (pending)"
    className="w-2 h-2 rounded-full bg-muted-foreground/30"
    testID="progress-step-3"
  />
  <View
    accessibilityLabel="Step 5 of 5 (pending)"
    className="w-2 h-2 rounded-full bg-muted-foreground/30"
    testID="progress-step-4"
  />
</View>
`;

exports[`ProgressIndicator Snapshot Testing should match snapshot with edge cases 1`] = `
<View
  className="flex-row justify-center items-center gap-2 "
  testID="progress-indicator-container"
>
  <View
    accessibilityLabel="Step 1 of 1 (current)"
    className="w-2 h-2 rounded-full bg-primary"
    testID="progress-step-0"
  />
</View>
`;
